### PR TITLE
marketplace with offer feature

### PR DESCRIPTION
**Pull request by psorinionut**

### CODE REVIEW

File: esdt-nft-marketplace/src/offer.rs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+                        "The auction does not contain the NFT"
+                    );
+                    require!(
+                        auction.current_bid == BigUint::zero(),
+                        "NFT auction has active bids"
+                    );
+                }
+                OptionalValue::None => sc_panic!("Must provide the auction id"),
+            };
+        }
```
Comment (dorin-iancu) [inline esdt-nft-marketplace/src/offer.rs:101]: A little less nesting makes the code clearer, i.e.:

`if &token_amount_in_marketplace < desired_amount { return; }`

`require!(opt_auction_id.is_some(), "Must provide the auction id");`

`let auction_id = unsafe { opt_auction_id.into_option().unwrap_unchecked(); }`
// your OptionalValue::Some() logic here

File: esdt-nft-marketplace/src/offer.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        let offer = Offer {
+            offer_token,
+            payment,
+            start_time: current_time,
+            deadline,
+            offer_owner: caller,
+        };
+
+        let offer_id = self.last_valid_offer_id().get() + 1;
+        self.last_valid_offer_id().set(offer_id);
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs]: Updated.

File: esdt-nft-marketplace/src/offer.rs (hunk 1 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs]: Updated.

File: esdt-nft-marketplace/src/offer.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        require!(!offer_mapper.is_empty(), "Offer does not exist");
+        offer_mapper.get()
+    }
+
+    #[view(getLastValidOfferId)]
+    #[storage_mapper("lastValidOfferId")]
+    fn last_valid_offer_id(&self) -> SingleValueMapper<u64>;
+
+    #[storage_mapper("offerById")]
+    fn offer_by_id(&self, offer_id: u64) -> SingleValueMapper<Offer<Self::Api>>;
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs:206]: I would keep the same implementation as it is for the auction.

**Review by dorin-iancu — APPROVED** [review]

**Review by CostinCarabas — APPROVED** [review]

**Pull request merged**